### PR TITLE
Kay and Laura.rb

### PR DESCRIPTION
**Pull request by lauracodecreations**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | method signature, body block, parameters
What are the advantages of using git when collaboratively working on one code base? | you make sure you are both working on the most updated version of the file.
What kind of relationship did you and your pair have with the unit tests? | we learned how to debug them, create them, and how to use pry with it, and their syntax
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | yes, it helped us to more efficiently iterate through the data structure. We used max_by to select max score and .sort_by to sort the strings by length and select, to select the key that matched the value
What was one method you and your pair used to debug code? | pry, and irb
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | how we improved our communication as the days passed, and how we became more efficient presenting and emerging ideas especially using the white board

### CODE REVIEW

File: Kay_and_Laura_adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
+require ‘csv’
+require ‘pry’
+
+def draw_letters
+
+ # all letters
+ letters = %W(a a a a a a a a a b b c c d d d d e e e e e e e e e e e e f f g g g h h i i i i i i i i i j k l l l l m m n n n n n n o o o o o o o o p p q r r r r r r s s s s t t t t t t u u u u v v w w x y y z )
```
Comment (CheezItMan) [inline Kay_and_Laura_adagrams.rb:7]: Good data structure to use!

File: Kay_and_Laura_adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ indices_used= []
+
+ # generate random index
+ index = Random.rand(min..max)
+
+ # get 10 unique indices, and saved them in indices used
+ 10.times do
+   # check if index is in indices_used array , if true generate new random index
+   while indices_used.include?(index)
+     index = Random.rand(min..max)
```
Comment (CheezItMan) [inline Kay_and_Laura_adagrams.rb:26]: Just a note, this loop could potentially go a long time.  You would do better to remove an element once you've selected it.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+def uses_available_letters?(input, letters_in_hand)
+  result = []
+  final_results = []
+  input.each_char do |letter|
+    # check if letter exists in letters_in_han
+    if letters_in_hand.include?(letter)
+      result << true
+    else
+      # user used a letter that does not exists
+      result << false
```
Comment (CheezItMan) [inline lib/adagrams.rb:47]: Maybe you should just return `false`.

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  return ten_letters
+end
+
+def uses_available_letters?(input, letters_in_hand)
+  result = []
+  final_results = []
+  input.each_char do |letter|
+    # check if letter exists in letters_in_han
+    if letters_in_hand.include?(letter)
+      result << true
```
Comment (CheezItMan) [inline lib/adagrams.rb:44]: What if a letter appears twice in `input`, but not in `letters_in_hand`

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    else
+      # user used a letter that does not exists
+      result << false
+    end
+  end
+  final_results << result.all? { |value| value == true}
+
+  input.each_char do |letter|
```
Comment (CheezItMan) [inline lib/adagrams.rb:52]: This seems to be doing what the above did, but better.  Avoid duplication of effort.

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+def score_word(word)
+  score = 0
+  group1 = %w(A E I O U L N R S T)
+  group2 = %w(D G)
+  group3 = %w(B C M P)
+  group4 = %w(F H V W Y)
+  group5 = %w(J X)
+  group6 = %w(Q Z)
+  word.upcase.each_char do |letter|
```
Comment (CheezItMan) [inline lib/adagrams.rb:73]: 👍

File: lib/adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    winner = ties_words[0]
+  else # ties
+    ties_words.each do |word|
+      if word.length == 10 # select word length 10 and return it as a winner
+        winner = word
+        # finds winner in data, and return it's value for the score
+        score = data.select{ |k, v| k == winner}.values[0]
+        return {:word => winner, :score => score}
+      else # select shortest word and return it as a winner
+        winner = ties_words.sort_by {|x| x.length}.first
```
Comment (CheezItMan) [inline lib/adagrams.rb:118]: You have a loop and inside that you're sorting each iteration.  This is a bit inefficient.

**Comment by CheezItMan** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |  Check
Both teammates contributed to the codebase |  I assume so
Small commits with meaningful commit messages |  Small commits, for your messages I suggest you avoid mentioning waves and talk about the functionality added.
**Code Requirements** |
`draw_letters` method | Check, but it's a bit awkward.  I left some notes in the codebase.
Uses appropriate data structure to store the letter distribution |  Well done!
All tests for `draw_letters` pass |  Check
`uses_available_letters?` method |  Check, but you have some redundant code here.  I left notes in the code.
All tests for `uses_available_letters?` pass | Check
`score_word` method |  Check
Uses appropriate data structure to store the letter scores | Nice!
All tests for `score_word` pass | Check
`highest_score_from` method |  Yes, but you have some inefficiencies here.  You loop through the tied words and each time the loop goes, it sorts the list again.  
Appropriately handles edge cases for tie-breaking logic |  It works.
All tests for `highest_score_from` pass | Check
**Overall** |  Overall you did ok, you had some inefficiencies and need to refactor your code.
